docs/provider-references.md

```diff
@@ -0,0 +1,352 @@
+    }
+  }
+}
+
+provider "awsname" {
+  region = "us-east-1"
+  alias = "default"
+}
+
+module "my_mod" {
+  source = "./mod"
+  providers = {
+    modaws = awsname.default
+  }
+}
+```
+
+```hcl
+# ./mod/mod.tf
+terraform {
+  required_providers {
+    modaws = {
+      source = "aws"
+      version = ">= 0.1"
+    }
+  }
+}
+
+
+resource "aws_s3_bucket" "foo" {
+  bucket_name = "foo"
+  provider = modaws
+}
+```
+
+The required_providers "source" field in the root module decomposed into `addrs.Provider{Type="aws", Namespace="hashicorp", Hostname="registry.opentofu.org"}`. As the default namespace is "hashicorp" and the default hostname is "registry.opentofu.org", we will continue to use the shorthand `addrs.Provider{Type="aws"}`. Next `addrs.LocalProviderConfig{LocalName: "aws_name"}` is created and within the root module maps to `addrs.Provider{Type="aws"}`. This provider local name is then used in all subsequent references within the the root module. The configuration is then mapped to `addrs.AbsProviderConfig{Module: Root, Provider: addrs.Provider{Type = "aws"}, Alias = "default"}` globally.
+
+The child module is passed the `addrs.AbsProviderConfig` and is internally referenceable within the module under `addrs.LocalProviderConfig{LocalName: "modaws"}`.
+
+Within that module, `addrs.LocalProviderConfig{LocalName: "modaws"}` now points at `addrs.Provider{Type = "aws"}` and is effectively replaced with `addrs.AbsProviderConfig{Module: Root, Provider: addrs.Provider{Type = "aws"}, Alias = "default"}` at runtime. This optimizes running as few provider instances as possible.
+
+If a new provider configuration were added to the module:
+```
+provider "modaws" {
+  region = "us-west-2"
+}
+```
+This would negate the override / deduplication above and result in `addrs.AbsProviderConfig{Module: MyMod, Provider: addrs.Provider{Type = "aws"}}`.
+
+#### Multiple Provider Aliases
+
+Multiple provider aliases can be supplied in required_providers via `configuration_aliases`. This requires that a caller of the module provide the requested aliases explicitly.
+
+Example:
+
+```
+terraform {
+  required_providers {
+    awsname = { # name added for clarity, usually Type == LocalName
+      source = "registry.opentofu.org/hashicorp/aws"
+      version = "0.124"
+    }
+  }
+}
+
+provider "awsname" {
+  region = "us-east-1"
+  alias = "default"
+}
+
+module "my_mod" {
+  source = "./mod"
+  providers = {
+    modaws.foo = awsname.default
+    modaws.bar = awsname
+  }
+}
+```
+
+```hcl
+# ./mod/mod.tf
+terraform {
+  required_providers {
+    modaws = {
+      source = "aws"
+      version = ">= 0.1"
+      configuration_aliases = [ modaws.foo, modaws.bar ]
+    }
+  }
+}
+```
+
+
+### Representation in State
+
+Resources in the state file note the `addrs.Provider` required to modify them. Due to the structure, all instances (for_each/count) of a resource must use the same provider.
+
+Note: This section should be expanded with examples.
```

Comment:
we should also mention that the json representation of the statefile is different to the actual statefile.

Reply:
cd6ba01ea5